### ISSUE TITLE
[Bug]: Fix flaky `Publish terraform plan` job in deployment and CI workflows

### ISSUE THREAD

**Issue by TylerHendrickson** [usdigitalresponse/usdr-gost#2557]
### Why is this issue important?

Visibility into Terraform plan output is a useful reviewable asset for pull requests and deployments, as it allows for close inspection of the impact that a code change will have on deployed infrastructure. However, failures related to Terraform plans should reflect a problem with the change itself, rather than a limitation of a workflow within our CI/CD pipelines.

#### Background context

`Publish terraform plan` is a [reusable workflow](https://docs.github.com/en/actions/using-workflows/reusing-workflows) implemented in `.github/workflows/publish-terraform-plan.yml`. It is called by a job (which has the same name) from each of the following workflows, rather than being executed as a standalone workflow:
- [Continuous Integration](https://github.com/usdigitalresponse/usdr-gost/actions/workflows/ci.yml)
- [Deploy to Staging](https://github.com/usdigitalresponse/usdr-gost/actions/workflows/deploy-staging.yml)
- [Deploy to Production](https://github.com/usdigitalresponse/usdr-gost/actions/workflows/deploy-production.yml)

Its purpose is to produce a markdown document that provides a summary of various Terraform-related commands executed during a prior step in the same workflow run. In particular, it provides a `diff`-formatted code block containing the output from `terraform plan`, which may then be reviewed by a PR author, PR reviewer, or (in the case of a Production deployment) a repo admin. The markdown document is then published in the following ways:
- **Always:** As a [job summary](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#adding-a-job-summary), which provided on the overview page for a specific workflow run ([example](https://github.com/usdigitalresponse/usdr-gost/actions/runs/7761364590#summary-21169802669)).
- **When run from a pull request's Continuous Integration workflow:** As a comment ([example](https://github.com/usdigitalresponse/usdr-gost/pull/2454#issuecomment-1888018472)), which is updated every time the workflow runs for the same PR.

### Current State

Especially large Terraform plan output causes the `Publish terraform plan` workflow to fail when running its `Reformat Plan` step, which produces the following output:

> ```Error: An error occurred trying to start process '/usr/bin/bash' with working directory '/home/runner/work/usdr-gost/usdr-gost'. Argument list too long```

#### Investigation note

The "argument list" mentioned in this error message appears (I have not fully verified this) to refer to the `$PLAN_RAW_OUTPUT` environment variable configured on the `Reformat Plan` step, and indicates that the value of the environment variable surpasses some max-length limitation imposed by GitHub Actions. It's worth noting that Google results for this error message often relate to issues with the [`github-script` action](https://github.com/actions/github-script), **but that is not the case here.** This may suggest that the GitHub Actions limitation applies not only to environment variables, but also action inputs. Various comments in this issue https://github.com/actions/runner/issues/1733 seem to confirm that this limitation does apply to environment variables, although I haven't been able to find any documentation that indicates the maximum allowed characters/bytes for an environment variable in GitHub Actions.

#### Security note

Given that the value of `$PLAN_RAW_OUTPUT` is set from the `tf-plan-output` workflow input, it may be tempting to resolve this issue by changing this (current code):
```yaml
- name: Reformat Plan
  run: |
    PLAN=$(echo "$PLAN_RAW_OUTPUT" | sed -E 's/^([[:space:]]+)([-+])/\2\1/g')
    echo "PLAN_REFORMATTED<<EOF" >> $GITHUB_ENV
    echo "$PLAN" >> $GITHUB_ENV
    echo "EOF" >> $GITHUB_ENV
  env:
    PLAN_RAW_OUTPUT: ${{ inputs.tf-plan-output }}
```
to something that sidesteps configuring any environment variable, like this:
```yaml
- name: Reformat Plan
  run: |
    PLAN=$(echo "${{ inputs.tf-plan-output }}" | sed -E 's/^([[:space:]]+)([-+])/\2\1/g')
    echo "PLAN_REFORMATTED<<EOF" >> $GITHUB_ENV
    echo "$PLAN" >> $GITHUB_ENV
    echo "EOF" >> $GITHUB_ENV
```
However, this is not recommended practice for inline scripts, as it creates a potential for malicious code execution (i.e. an injection attack). See [this documentation](https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-an-intermediate-environment-variable) for more details.

### Expected State

- When the `Reformat Plan` step fails, the overall job/workflow does not. 
- The plan is still available for review in some fashion.

### Implementation Plan

I'm definitely open to suggestions, but to offer one possible alternative, the behavior of `Publish terraform plan` could be modified as-follows:
1. Before running the `Reformat Plan` step, add a new step that uploads the contents of the `tf-plan-output` [workflow input parameter](https://docs.github.com/en/actions/using-workflows/reusing-workflows#using-inputs-and-secrets-in-a-reusable-workflow) (which is referenced within the workflow body as `inputs.tf-plan-output`) as a job artifact. This would ensure that the plan is always available to review in some fashion.
2. Modify the `Reformat Plan` step in the following way:
    - Give the step an identifier, e.g. `id: reformat-plan`
    - Configure the step with `continue-on-error: true`
3. Add a new [conditional step](https://docs.github.com/en/actions/using-jobs/using-conditions-to-control-job-execution) _after_ `Reformat plan`, which only executes if `Reformat plan` failed, e.g. 
  ```if: steps.reformat-plan.outcome == 'failure'```
  This step would be responsible for setting an alternate value for the `$PLAN_REFORMATTED` environment variable (which is normally set by `Reformat Plan`). The alternate value could be a message stating that the plan was too large to include and provide a download link (or instruct a user on how to find the link) for the artifact that was uploaded (see the first item in this list).

**Comment by TylerHendrickson** [comment]
Some occurrences for this bug:
- [`Continuous Integration` workflow run](https://github.com/usdigitalresponse/usdr-gost/actions/runs/7753343129) for  https://github.com/usdigitalresponse/usdr-gost/pull/2555
- [`Deploy to Production` workflow run](https://github.com/usdigitalresponse/usdr-gost/actions/runs/7669715024) for [Release v2024.9](https://github.com/usdigitalresponse/usdr-gost/releases/tag/release%2F2024.9)

**Comment by TylerHendrickson** [comment]
Fixed by #2966
